views/js/portableElementRegistry/factory/factory.js

```diff
@@ -0,0 +1,200 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2016 (original work) Open Assessment Technologies SA;
+ *
+ */
+define(['lodash', 'core/promise', 'core/eventifier'], function (_, Promise, eventifier){
+    'use strict';
+
+    var _requirejs = window.require;
+
+    return function portableElementRegistry(methods){
+
+        var _loaded = false;
+        var __providers = {};
+
+        return eventifier(_.defaults(methods || {}, {
+            _registry : {},
+            get : function get(typeIdentifier, version){
+
+                if(this._registry[typeIdentifier]){
+                    //check version
+                    if(version){
+                        return _.find(this._registry[typeIdentifier], version);
+                    }else{
+                        //latest
+                        return _.last(this._registry[typeIdentifier]);
+                    }
+                }
+            },
+            registerProvider : function registerProvider(moduleName){
+                __providers[moduleName] = null;
+                return this;
+            },
+            resetProviders : function resetProviders(){
+                __providers = {};
+                return this;
+            },
+            loadProviders : function loadProviders(callback){
+                var self = this;
+                var providerLoadingStack = [];
+                _.forIn(__providers, function(provider, id){
+                    if(provider === null){
+                        providerLoadingStack.push(id);
+                    }
+                });
+                _requirejs(providerLoadingStack, function(){
+                    _.each([].slice.call(arguments), function(provider){
+                        if(provider && _.isFunction(provider.load)){
+                            __providers[providerLoadingStack.shift()] = provider;
+                        }
+                    });
+                    callback();
+                    self.trigger('providersloaded');
+                });
+            },
+            getAllVersions : function getAllVersions(){
+                var all = {};
+                _.forIn(this._registry, function (versions, id){
+                    all[id] = _.map(versions, 'version');
+                });
+                return all;
+            },
+            getRuntime : function getRuntime(typeIdentifier, version){
+                var pci = this.get(typeIdentifier, version);
+                if(pci){
+                    return _.assign(pci.runtime, {
+                        id : pci.typeIdentifier,
+                        label : pci.label,
+                        baseUrl : pci.baseUrl
+                    });
+                }else{
+                    this.trigger('error', 'no portable element runtime found' , typeIdentifier, version);
```

Comment:
Should it not be better to provide a single argument with error events ?
This way we could use an unified version of error handling.
Also Promises cannot return more than one argument.